### ISSUE TITLE
[Bug] PaypalButtons Event handlers hold stale state

### ISSUE THREAD

**Issue by wesbos** [paypal/paypal-js#590]
With react-paypal-js (`"@paypal/react-paypal-js": "^8.7.0"`), the event handlers seem to create a closure and hold old state - its not clear from the docs on how to avoid this, or update.

An example:

```jsx
function Paypal() {
  const [count, setCount] = useState(0);
  return (
    <>
      <button onClick={() => setCount(count + 1)}>Count: {count}</button>
      <PayPalButtons
        onClick={() => {
          console.log(count) // Always 0
        }}
      />
    </>
  );
}
```

Click the button a few times, state updates, but when any of the event handlers of PaypalButton are called, it's the stale state.

Seems you are creating a closure of the event handler functions. 

This is a problem because people 

1. go to my checkout, enter their info (which updates cart state)
2. click Paypal as a payment option - this renders the PaypalButtons component
3. Then they go back to their checkout and change things like coupon, name or email address.
4. They click the paypal button

So the state is stale back to when the button was initially rendered. 

Am I supposed to force a re-render of the button every time my state changes? 

I'd consider this a bug as I'd expect the onClick to just use the state, but if it's not a bug, it needs documentation as it doesn't behave as React event handlers do.

**Comment by wesbos** [comment]
Seems I can reload the entire thing like this:

```jsx
const [{ options }, dispatch] = usePayPalScriptReducer();
// then whenever any state changes
dispatch({ type: 'resetOptions', value: { ...options } });
```

This can't be the fix though, right?

**Comment by wesbos** [comment]
Apologies for the multiple comments, I see there is also a `forceReRender` prop for the button, which I assume is like a useEffect dependency

This works, but has the downside of the button flickering and animating any time state of changed.

https://github.com/user-attachments/assets/81bac2e7-486a-42f7-909a-ebb2387464ba

And if this happens too quickly (like state is updated on keystroke), these errors appear in the console:

<img width="480" alt="Screenshot 2024-12-04 at 1 25 10 PM" src="https://github.com/user-attachments/assets/b2903b52-9123-425c-a350-2fb8d31b90c3">

I just want to use state in my `onError, onCancel, onApprove, onClick, createOrder` methods without having to jump through hoops

**Comment by Dandrum** [comment]
Hi there,

I'm using nextjs 14.* and have exactly the same problem. It seems that the SDK is not made for SSR because by default every next component is rendered on the server and then hydrated therefore the onFunctions dont get the current state.

The button component needs a 'use client' tag for use in next or a client-only solution. 

Do you have the animation problem with the dispatch too, the forceRerender with the animation is not a nice solution.

**Comment by wesbos** [comment]
yep, force re render causes the animation. Frustrating that we haven't heard from anyone about this issue - I've DM'd with a few Dev Rel folks at paypal but not a reply here. 

@gregjopa @nbierdeman @danzhaaspaypal @elizabethmv @sebastianfdz  Apologies for the tag - any way you can poke someone who can help here?

**Comment by gregjopa** [comment]
Hey @wesbos, apologies for the delay. The `forceReRender` prop was designed to reload the JS SDK script when certain query param values change that directly affect button render eligibility like `currency` and `intent`. It technically works for solving the problem you ran into when updating the order amount with the `createOrder()` callback but provides a bad buyer experience since the buttons iframe reloads which triggers the button label animation. One idea for a short-term solution is to leverage a React class component to wrap the provider and buttons components so the callbacks can be referenced with `this`. It's mentioned in this stackoverflow post: https://stackoverflow.com/questions/69259922/how-to-pass-dynamic-amount-to-paypal-react-sdk-button-without-re-render. Would that solution work for your integration?

Unfortunately when we first created this library, we didn't know react well enough to understand how to best pass functions as props. It should be easy for the `createOrder()` function definition to change based on the buyer's choices without re-rendering the component since that is supported when doing a vanilla JS integration with the PayPal JS SDK. It's common for merchants to support adding/removing items from the buyer's shopping cart or applying a discount code which change the order payload. 

One breaking change idea we have discussed in the past is leveraging [useCallback](https://react.dev/reference/react/useCallback) to wrap the various callbacks required by a PayPal Buttons integration. The thought is we would teach merchant developers to wrap every callback with this and be responsible for controlling the state changes. Here's a rough example:
```react
<PayPalButtons createOrder={useCallback(merchantsCreateOrderCallback)} />
```

I have not tested this idea yet. We are definitely open to other interface changes too so we can better handle this use case.

**Comment by wesbos** [comment]
Thanks for the reply @gregjopa!

I do like the callback approach, but I think if you were to just make it work like normal react, where the function is called when the event happens — its kind of what we would expect. If users need to cache the function between renders, they can use useCallback.

If you don't want to break the API, could you provide some sort of useCallback-like function that we use to wrap our functions that would get the dynamic data at the time of the event?


```ts
<PayPalButtons createOrder={reactive(merchantsCreateOrderCallback)} />
```

**Comment by imbrian** [comment]
Hey @wesbos, just spoke about this with @gregjopa. You're right about the cause; the callbacks are only passed once when the SDK component is initially rendered, creating the stale closure.

This is something we'd been tracking, and we have a proposed solution to observe callback updates without an API change _and_ without re-rendering the SDK components. 

In the meantime, tracking state via `useRef` (or an external store, state container, etc) could offer a (hacky) workaround...
```js
function Paypal() {
  const count = useRef(0);
  const [,forceUpdate] = useState({})
  const setCount = (cb) => {
    count.current = cb(count.current)
    forceUpdate({})
  }
  
  return (
    <>
      <button onClick={() => setCount(prev => prev + 1)}>Count: {count.current}</button>
      <PayPalButtons
        onClick={() => {
          console.log(count.current)
        }}
      />
    </>
  );
}
```

**Comment by imbrian** [comment]
@wesbos, a fix for this will be in v8.8.0 when #606 is merged.

**Comment by wsbrunson** [comment]
`@paypal/react-paypal-js@8.8.1` is now available

**Comment by gregjopa** [comment]
Thanks @imbrian and @wsbrunson for getting this fix out! It's awesome that we didn't need any interface changes to get the `createOrder()` and `onApprove()` function props to work as expected.

**Comment by wesbos** [comment]
amazing! Thank you all - this is great to hear. Nice fix as well without changing the API.

**Comment by m345054** [comment]
I'm using v8.8.2, and it looks like the createOrder function still isn't fixed. Is anyone else experiencing this?

```
  const createOrder = useCallback(async () => {
    console.log("payment", paymentSession)
    
    if (!paymentSession) {
      return Promise.reject(t("checkout.payment.no_session"))
    }
    return paymentSession.data.id as string
  }, [paymentSession, t])
```

**Comment by imbrian** [comment]
@m345054, you're right; this fix was only for the Buttons component. We do have a fix for CardFields planned.
